core/services/ocr2/plugins/ocr2keeper/evm21/encoder.go

```diff
@@ -159,9 +174,49 @@ func (enc EVMAutomationEncoder21) KeysFromReport(b []byte) ([]ocr2keepers.Upkeep
 	return keys, nil
 }
 
+// Extract the plugin will call this function to accept/transmit reports
+func (enc EVMAutomationEncoder21) Extract(raw []byte) ([]ocr2keepers.ReportedUpkeep, error) {
+	report, err := enc.packer.UnpackReport(raw)
+	if err != nil {
+		return nil, err
+	}
+	reportedUpkeeps := make([]ocr2keepers.ReportedUpkeep, len(report.UpkeepIds))
+	for i, upkeepId := range report.UpkeepIds {
+		triggerW, err := enc.packer.UnpackTrigger(upkeepId, report.Triggers[i])
+		if err != nil {
+			// TODO: log error and continue instead?
+			return nil, fmt.Errorf("%w: failed to unpack trigger", err)
+		}
+		logExt := logprovider.LogTriggerExtension{}
+
+		switch getUpkeepType(upkeepId.Bytes()) {
+		case logTrigger:
+			logExt.TxHash = common.BytesToHash(triggerW.TxHash[:]).Hex()
+			logExt.LogIndex = int64(triggerW.LogIndex)
+		default:
+		}
+		trigger := ocr2keepers.NewTrigger(
+			int64(triggerW.BlockNum),
+			string(triggerW.BlockHash[:]),
+			logExt,
+		)
+		payload := ocr2keepers.NewUpkeepPayload(
+			upkeepId,
+			int(logTrigger),
+			"",
+			trigger,
+			[]byte{},
+		)
+		reportedUpkeeps[i] = ocr2keepers.ReportedUpkeep{
```

Comment:
do we need to populate other fields in this struct?

Reply:
We really do need upkeep id and trigger data. And @amirylm now that I look at this again, we don't need extra fields as the trigger data will be all that is necessary for the check block that we talked about.